### PR TITLE
HIPP-1615: Refactor to handle APIM deployment responses failures

### PR DESCRIPTION
**Pull request by victorarbuesmallada**
APIM outages would prevent our users to see deployment statuses. Instead of propagating the outage errors, we are just returning different deployment status types.

### CODE REVIEW

File: app/uk/gov/hmrc/apihubapplications/controllers/DeploymentsController.scala [hunk 1]
```diff
@@ -70,21 +70,10 @@ class DeploymentsController @Inject()(
...
-      }
-  }
-
-  private def getOasVersion(response: Option[DeploymentResponse]): Option[String] = response match {
-    case Some(SuccessfulDeploymentResponse(_, oasVersion)) => Some(oasVersion)
-    case _ => None
+      deploymentsService.getDeployments(publisherRef)
+        .map(response =>
+          Ok(Json.toJson(DeploymentStatuses(response)))
+        )
```
Comment (PaulCDurham) [inline app/uk/gov/hmrc/apihubapplications/controllers/DeploymentsController.scala:76]: What happens to the Left result? Everywhere else we have a `case Left(e) => throw e` that handles that explicitly. Does Play know what to do with the Left?

**Review by PaulCDurham — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/apihubapplications/controllers/DeploymentsController.scala (hunk 1 above)
Comment (victorarbuesmallada) [reply app/uk/gov/hmrc/apihubapplications/controllers/DeploymentsController.scala:76]: At this point, I thought we were returning just a sequence of deployment statuses and the left result is handled on the service, which turns that into an `Unexpected` deployment status

File: app/uk/gov/hmrc/apihubapplications/connectors/APIMConnectorImpl.scala [hunk 2]
```diff
@@ -188,9 +189,14 @@ class APIMConnectorImpl @Inject()(
...
           logger.warn(s"Received 403 back from APIM ${environment.toString} whilst looking up publisher reference: $publisherReference and useProxyForSecondary: $useProxyForSecondary. Full url: ${url.toString}")
-          Right(None)
+          Left(raiseApimException.unexpectedResponse(e.statusCode, context))
         case Left(e) => Left(raiseApimException.unexpectedResponse(e.statusCode, context))
       }
+      . recover {
+        case NonFatal(e) =>
+          logger.warn(s"Error occurred while retrieving APIM deployments", e)
+          Left(raiseApimException.unexpectedResponse(500, context))
+      }
```
[the remark this answers is not in the document]
Comment (victorarbuesmallada) [reply app/uk/gov/hmrc/apihubapplications/connectors/APIMConnectorImpl.scala:199]: the [http client](https://github.com/hmrc/http-verbs/blob/2710cda3e057ab5c99c6d9210399d1364f503bad/http-verbs-play-30/src/main/scala/uk/gov/hmrc/http/client/HttpClientV2Impl.scala#L309-L312) that we use makes the future fail in timeout/gateway issues, that's why I needed the `recover` block.

File: app/uk/gov/hmrc/apihubapplications/services/DeploymentsService.scala [hunk 3]
```diff
@@ -69,8 +75,26 @@ class DeploymentsService @Inject()(
   def getDeployment(
                      publisherRef: String,
                      environmentName: EnvironmentName
-                   )(implicit hc: HeaderCarrier): Future[Either[ApimException, Option[DeploymentResponse]]] = {
-    apimConnector.getDeployment(publisherRef, environmentName)
+                   )(implicit hc: HeaderCarrier): Future[Either[ApimException, Option[DeploymentResponse]]] =
+    EitherT(apimConnector.getDeployment(publisherRef, environmentName))
+      .leftSemiflatTap(
+        exception => Future.successful(logger.warn(customUnknownDeploymentStatusMessage, exception))
+      ).value
```
Comment (PaulCDurham) [inline app/uk/gov/hmrc/apihubapplications/services/DeploymentsService.scala:82]: Why has getDeployment changed? In fact I don't think we need it. This was used by the controller but has been replaced by getDeployments. I'd just delete it.

**Review by PaulCDurham — CHANGES_REQUESTED** [review]
Can you check if it's used (I don't think it is) and remove getDeployment from DeploymentsService?

**Review by PaulCDurham — APPROVED** [review]

**Pull request merged**